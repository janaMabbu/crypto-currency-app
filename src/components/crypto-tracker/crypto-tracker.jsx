import React, { PureComponent, Fragment } from 'react'
import { connect } from 'react-redux'
import ImmutablePropTypes from 'react-immutable-proptypes'
import PropTypes from 'prop-types'
import { getCurrencies, loadCurrencies } from 'ducks/get-crypto-currencies'
import { formatCurrency } from 'ducks/helpers'

import './crypto-tracker.less'

export class CryptoTracker extends PureComponent {
  static propTypes = {
    currencies: ImmutablePropTypes.list,
    loadCurrencies: PropTypes.func.isRequired
  }

  componentDidMount () {
    const { loadCurrencies } = this.props
    // load the currencies in this array
    loadCurrencies(['btc','bch','ltc', 'usdt', 'jana'])
    // scrolling to top on browser navigation
    window.scrollTo(0, 0)
  
  }


  setRoutePath = (path) => {
    // this navigated the user to next page
    this.props.router.push(`cryptotracker/${path}`)
  }


  render () {
    const { currencies } = this.props
    return (
      <div className="row crypto-app">
        <div className="container crypto-app__tbody">
  <div className="row crypto-app__t-row">
    <div className="col crypto-app__t-col">
      <strong>Coin Name</strong>
    </div>
    <div className="col crypto-app__t-col">
      <strong>Symbol</strong>
    </div>
    <div className="col crypto-app__t-col">
      <strong>Latest Price</strong>
    </div>
    <div className="col crypto-app__t-col">
    </div>
  </div>
    { currencies.map((coin)=> {
          if(coin.get('status', false)) {
            return this.renderSuccessRow(coin)
          } else if(!coin.get('status', true)) {
            return this.renderFailureRow(coin)
          } else {
            return (
              <div key={coin.get('id')} className="row crypto-app__t-row crypto-app__loading">
                <div className="spinner-border"></div>
              </div>)
          }
        }) }
  </div>
</div>
    )
  }

  renderSuccessRow = (coin) => {
    const id = coin.get('id')
     return(
    <div className="row crypto-app__t-row" key={coin.get('id')}>
      <div className="col-3 crypto-app__t-col">{coin.getIn(['data', 'coin_name'], '')}</div>
      <div className="col-3 crypto-app__t-col">{coin.getIn(['data', 'coin_id'], '')}</div>
      <div className="col-3 crypto-app__t-col">{formatCurrency(coin.getIn(['data', 'last_price'], 0),2)}</div>
      <div className="col-sm-3 crypto-app__t-col"><button className="container crypto-app__button" id={ id } onClick={()=>this.setRoutePath(coin.get('id')) }>Details</button></div>
    </div>
    )
  }

  renderFailureRow = (coin) => {
     return(
      <div className="row crypto-app__t-row crypto-app__error" key={coin.get('id')}>
        something wrong with loading coin with id "{coin.get('id')}"
      </div>
    )
  }

}
const mapStateToProps = state => {
  return {
    currencies: getCurrencies(state)
  }
}
const mapDispatchToProps = {
  loadCurrencies
}
export default connect(mapStateToProps, mapDispatchToProps)(CryptoTracker)
