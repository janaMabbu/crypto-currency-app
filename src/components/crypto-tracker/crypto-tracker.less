@import "~less/common.less";

.crypto-app {
  flex-direction: column;
  align-items: center;
  padding: 50px;
  min-height: 500px;

  @media screen and (max-width: 576px) {
    padding: 15px;
    min-height: 300px;
  }

  &__error {
    color: red;
    justify-content: center;
  }

  &__loading {
    text-align: center;
    padding: 5px;
  }

  &__content1 {
    text-align: center;
    margin: 30px;
  }
}
