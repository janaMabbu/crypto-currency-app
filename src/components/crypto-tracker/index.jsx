export default from './crypto-tracker'