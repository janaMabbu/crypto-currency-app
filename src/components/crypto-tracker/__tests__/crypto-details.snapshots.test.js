import { shallow } from 'enzyme'
import React from 'react'
import Immutable from 'immutable'
import { CryptoTracker } from 'components/crypto-tracker/crypto-tracker'

describe('Component - CryptoTracker', () => {
  let props
  beforeEach(() => {
    window.scrollTo = jest.fn()
    props = {
      currencies: Immutable.fromJS([{
        vol_24hr_pcnt:'-27.21',
        last_price:'9705.18401230',
        success:true,
        utc_date:'2020-02-22 21:41:01',
        coin_id:'BTC',
        currency:'USD',
        volume_24hr:'4126767296',
        time_stamp:'1582407661',
        price_24hr_pcnt:'0.50',
        currency_name:'United States Dollar',
        coin_name:'Bitcoin',
        source:'BraveNewCoin'
      },
      {
        vol_24hr_pcnt:'-27.21',
        last_price:'9705.18401230',
        success:true,
        utc_date:'2020-02-22 21:41:01',
        coin_id:'BhC',
        currency:'USD',
        volume_24hr:'4126767296',
        time_stamp:'1582407661',
        price_24hr_pcnt:'0.50',
        currency_name:'United States Dollar',
        coin_name:'Bitcoin',
        source:'BraveNewCoin'
      }]),
      loadCurrencies: jest.fn()
    }
  })
  it('should render the CryptoTracker snapshot', () => {
    const wrapper = shallow(<CryptoTracker {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should only display only success rows', () => {
    const wrapper = shallow(<CryptoTracker {...props}/>)
    expect(wrapper.find('.crypto-app__error').length).toBe(0)
    expect(wrapper.find('.crypto-app__t-row').length).toBe(3)
  })

  it('should display one failure row, one header, one success row', () => {
    const wrapper = shallow(<CryptoTracker {...props}/>)
    wrapper.setProps({ currencies: props.currencies.setIn([0,'status'], false) })
    expect(wrapper.find('.crypto-app__error').length).toBe(1)
    expect(wrapper.find('.crypto-app__t-row').length).toBe(3)
  })

  it('should render the CryptoTracker fallback text message', () => {
    const wrapper = shallow(<CryptoTracker {...props} />)
    wrapper.setProps({ currencies: props.currencies.setIn([0,'status'], false) })
    expect(wrapper.contains('something went wrong loading the coin!')).toBe(true)
  })
})