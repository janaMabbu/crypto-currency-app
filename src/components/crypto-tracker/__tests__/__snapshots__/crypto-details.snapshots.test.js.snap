// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - CryptoTracker should render the CryptoTracker snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CryptoTracker
    currencies={
      Immutable.List [
        Immutable.Map {
          "vol_24hr_pcnt": "-27.21",
          "last_price": "9705.18401230",
          "success": true,
          "utc_date": "2020-02-22 21:41:01",
          "coin_id": "BTC",
          "currency": "USD",
          "volume_24hr": "4126767296",
          "time_stamp": "1582407661",
          "price_24hr_pcnt": "0.50",
          "currency_name": "United States Dollar",
          "coin_name": "Bitcoin",
          "source": "BraveNewCoin",
        },
        Immutable.Map {
          "vol_24hr_pcnt": "-27.21",
          "last_price": "9705.18401230",
          "success": true,
          "utc_date": "2020-02-22 21:41:01",
          "coin_id": "BhC",
          "currency": "USD",
          "volume_24hr": "4126767296",
          "time_stamp": "1582407661",
          "price_24hr_pcnt": "0.50",
          "currency_name": "United States Dollar",
          "coin_name": "Bitcoin",
          "source": "BraveNewCoin",
        },
      ]
    }
    loadCurrencies={
      [MockFunction] {
        "calls": Array [
          Array [
            Array [
              "btc",
              "bch",
              "ltc",
              "usdt",
              "jana",
            ],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="container crypto-app__tbody"
      >
        <div
          className="row crypto-app__t-row"
        >
          <div
            className="col crypto-app__t-col"
          >
            <strong>
              Coin Name
            </strong>
          </div>
          <div
            className="col crypto-app__t-col"
          >
            <strong>
              Symbol
            </strong>
          </div>
          <div
            className="col crypto-app__t-col"
          >
            <strong>
              Latest Price
            </strong>
          </div>
          <div
            className="col crypto-app__t-col"
          />
        </div>
        Immutable.List [
          <div
            className="row crypto-app__t-row crypto-app__loading"
          >
            <div
              className="spinner-border"
            />
          </div>,
          <div
            className="row crypto-app__t-row crypto-app__loading"
          >
            <div
              className="spinner-border"
            />
          </div>,
        ]
      </div>,
      "className": "row crypto-app",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row crypto-app__t-row"
          >
            <div
              className="col crypto-app__t-col"
            >
              <strong>
                Coin Name
              </strong>
            </div>
            <div
              className="col crypto-app__t-col"
            >
              <strong>
                Symbol
              </strong>
            </div>
            <div
              className="col crypto-app__t-col"
            >
              <strong>
                Latest Price
              </strong>
            </div>
            <div
              className="col crypto-app__t-col"
            />
          </div>,
          Immutable.List [
            <div
              className="row crypto-app__t-row crypto-app__loading"
            >
              <div
                className="spinner-border"
              />
            </div>,
            <div
              className="row crypto-app__t-row crypto-app__loading"
            >
              <div
                className="spinner-border"
              />
            </div>,
          ],
        ],
        "className": "container crypto-app__tbody",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col crypto-app__t-col"
              >
                <strong>
                  Coin Name
                </strong>
              </div>,
              <div
                className="col crypto-app__t-col"
              >
                <strong>
                  Symbol
                </strong>
              </div>,
              <div
                className="col crypto-app__t-col"
              >
                <strong>
                  Latest Price
                </strong>
              </div>,
              <div
                className="col crypto-app__t-col"
              />,
            ],
            "className": "row crypto-app__t-row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <strong>
                  Coin Name
                </strong>,
                "className": "col crypto-app__t-col",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Coin Name",
                },
                "ref": null,
                "rendered": "Coin Name",
                "type": "strong",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <strong>
                  Symbol
                </strong>,
                "className": "col crypto-app__t-col",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Symbol",
                },
                "ref": null,
                "rendered": "Symbol",
                "type": "strong",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <strong>
                  Latest Price
                </strong>,
                "className": "col crypto-app__t-col",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Latest Price",
                },
                "ref": null,
                "rendered": "Latest Price",
                "type": "strong",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "col crypto-app__t-col",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="spinner-border"
            />,
            "className": "row crypto-app__t-row crypto-app__loading",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "spinner-border",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="spinner-border"
            />,
            "className": "row crypto-app__t-row crypto-app__loading",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "spinner-border",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container crypto-app__tbody"
        >
          <div
            className="row crypto-app__t-row"
          >
            <div
              className="col crypto-app__t-col"
            >
              <strong>
                Coin Name
              </strong>
            </div>
            <div
              className="col crypto-app__t-col"
            >
              <strong>
                Symbol
              </strong>
            </div>
            <div
              className="col crypto-app__t-col"
            >
              <strong>
                Latest Price
              </strong>
            </div>
            <div
              className="col crypto-app__t-col"
            />
          </div>
          Immutable.List [
            <div
              className="row crypto-app__t-row crypto-app__loading"
            >
              <div
                className="spinner-border"
              />
            </div>,
            <div
              className="row crypto-app__t-row crypto-app__loading"
            >
              <div
                className="spinner-border"
              />
            </div>,
          ]
        </div>,
        "className": "row crypto-app",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="row crypto-app__t-row"
            >
              <div
                className="col crypto-app__t-col"
              >
                <strong>
                  Coin Name
                </strong>
              </div>
              <div
                className="col crypto-app__t-col"
              >
                <strong>
                  Symbol
                </strong>
              </div>
              <div
                className="col crypto-app__t-col"
              >
                <strong>
                  Latest Price
                </strong>
              </div>
              <div
                className="col crypto-app__t-col"
              />
            </div>,
            Immutable.List [
              <div
                className="row crypto-app__t-row crypto-app__loading"
              >
                <div
                  className="spinner-border"
                />
              </div>,
              <div
                className="row crypto-app__t-row crypto-app__loading"
              >
                <div
                  className="spinner-border"
                />
              </div>,
            ],
          ],
          "className": "container crypto-app__tbody",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="col crypto-app__t-col"
                >
                  <strong>
                    Coin Name
                  </strong>
                </div>,
                <div
                  className="col crypto-app__t-col"
                >
                  <strong>
                    Symbol
                  </strong>
                </div>,
                <div
                  className="col crypto-app__t-col"
                >
                  <strong>
                    Latest Price
                  </strong>
                </div>,
                <div
                  className="col crypto-app__t-col"
                />,
              ],
              "className": "row crypto-app__t-row",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <strong>
                    Coin Name
                  </strong>,
                  "className": "col crypto-app__t-col",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Coin Name",
                  },
                  "ref": null,
                  "rendered": "Coin Name",
                  "type": "strong",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <strong>
                    Symbol
                  </strong>,
                  "className": "col crypto-app__t-col",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Symbol",
                  },
                  "ref": null,
                  "rendered": "Symbol",
                  "type": "strong",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <strong>
                    Latest Price
                  </strong>,
                  "className": "col crypto-app__t-col",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Latest Price",
                  },
                  "ref": null,
                  "rendered": "Latest Price",
                  "type": "strong",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "col crypto-app__t-col",
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="spinner-border"
              />,
              "className": "row crypto-app__t-row crypto-app__loading",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "spinner-border",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="spinner-border"
              />,
              "className": "row crypto-app__t-row crypto-app__loading",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "spinner-border",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
